### DISCUSSION TITLE
Add possibility to delay between requests

### DISCUSSION THREAD

**Discussion by okainov** [renovatebot/renovate#11219] [Request Help]
### What would you like Renovate to be able to do?

We've stuck stumbled upon Github.com blocking our renovate job due to [secondary rate limit](https://docs.github.com/en/rest/overview/resources-in-the-rest-api#secondary-rate-limits). I was looking for some configuration option to add some delay between requests (like 0.5-1 seconds), but seems like there is none.

### If you have any ideas on how this should be implemented, please tell us here.

-

### Is this a feature you are interested in implementing yourself?

No

**Comment by viceice** [comment]
Renovate is respecting the github rate-limit headers, so you must have some bad configuration to hit rate limts.

So please provide your configuration and your debug log parts, where you see the errors with some sourrounding context lines.

**Reply by okainov** [reply]
We have ~100 repositories with different settings. Also the same API key is used in other places, like our regular CI

**Reply by viceice** [reply]
The hosted renovate app is running on thousands of repos and didn't hit  any issues. So whats your bot configuration?

You should not run renovate more than hourly. Maybe you should switch to free renovate-onn-prem.

https://github.com/whitesource/renovate-on-prem

**Reply by okainov** [reply]
Sorry, are you reading what I'm saying? We do run it on prem.

We use `renovate/renovate:latest` Docker image and this command line. That's all "configuration". Per-repo configurations are different, as I wrote, we have ~100 repos (and ~10k repos in our private org in total)
```
                String cmd = "renovate" +
                             " --token ${GITHUB_TOKEN}" +
                            // onboarding false would lead to simply silently skipping repos without config,
                            // so we avoid extra noise to non-participating repos
                             " --onboarding false" +
                             " --semantic-commits enabled" +
                             " --labels=renovate,deps" +
                             // Create Github issue with the summary of Renovate PRs
                             " --dependency-dashboard true" +
                             " --autodiscover" +
                             " --autodiscover-filter '*<our-expressions>*'"
```

**Reply by rarkins** [reply]
> Sorry, are you reading what I'm saying? We do run it on prem.

He's referring to the product called "WhiteSource Renovate On-Premises" which he then linked to, and I assume you did not click on.

> We use `renovate/renovate:latest` Docker image 

^^ not this

WhiteSource Renovate On-Prem works as an app, and I recommend you consider it if you're hitting rate limits running with a regular user token.

Please try to keep your tone polite and respectful if participating in this repository.

**Reply by okainov** [reply]
I did click there =) Probably the description is a bit not clear, because I saw "Docker hub", "renovate" and got a conclusion that it's the same.

Anyway, we're quite fine with the current docker image, today it works fine again, just yesterday there were two times the run has failed due to this secondary rate limit. The only ask was to have some kind of delay which would probably work with the cost of a bit increased run time.

**Reply by rarkins** [reply]
In general this is not something we'd *want* to do unless it's simple. It would seem that it's being applied to "real" user accounts and not bots/apps.

https://github.com/renovatebot/renovate/issues/6359 is perhaps what you're asking for, but I don't think it reliably solves it. I think the secondary rate limit is likely triggered by too many PRs created too rapidly, whereas http rate limiting will take effect on *every* request.

What you most likely need is:
- at http layer: rate limiting PUT/POST requests instead of all, or
- at application layer: rate limiting PR creation specifically. but we don't know what their limits are or how to read them at runtime

**Reply by malmor** [reply]
In case you are wondering how the "secondary rate limit" might be triggered:

We started using `RENOVATE_PAGINATE_ALL` (as mentioned in https://github.com/renovatebot/renovate/issues/8852#issuecomment-906227536, running renovate self-hosted) while updating ~100 Repositories.

Some of them are using dependencies like https://github.com/aws/aws-sdk-go-v2 or https://github.com/microsoft/fluentui, both containing 5000+ GitHub releases - this results in quite a few requests against the GitHub Releases API 😉.

**Reply by okainov** [reply]
Still there, actually now it became much worse, months ago it was failing like ~once in couple of days, now it's successful ~once in couple of days.... The problem is I cannot even simply rerun renovate, because it would be probably stuck again and/or will go again over the repos which were renovated before. Given that Github does send in the headers `retry-after` amount of seconds one needs to wait before retrying, I would really expect Renovate to read this and actually try to retry again instead of simply failing the whole batch.

**Reply by okainov** [reply]
@malmor  this sounds like something I might try, thanks. Now it became completely bad, not a single successfull run in couple of days...

@rarkins I've got an information from our team who were in contact with Github that secondary rate limit is NOT shared between different tokens of the same user. And couple of weeks ago I have generated dedicated token for Renovate exclusively (for the same user as our main API activities) and it was still failing hitting secondary rate limit. Here is the log, now I'm quite sure Renovate does NOT respect rate limit nor respects `retry-after` header:
![image](https://user-images.githubusercontent.com/918446/136262514-c22d49ee-6dd5-4908-8738-5ef35c06f6d6.png)

**Reply by rarkins** [reply]
The log message shows nothing except Renovate has reached the rate limit, not ignored it. Here are your choices:
- Use the hosted github.com app
- Use WhiteSource Renovate On-Premises
- Contribute to this repo if you can fix it

**Reply by felipecrs** [reply]
Another interesting option is https://github.com/app-sre/github-mirror. But not sure how it compares against native redis support within Renovate.
